src/element/annotations.rs

```diff
@@ -0,0 +1,110 @@
+use crate::Symbol;
+
+/// An ordered sequence of symbols that convey additional, application-specific information about
+/// their associated Ion value.
+/// ```
+/// use ion_rs::element::Annotations;
+/// let annotations: Annotations = ["foo", "bar", "baz"].into();
+/// for annotation in &annotations {
+///     assert_eq!(annotation.text().map(|s| s.len()), Some(3));
+/// }
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Annotations {
+    symbols: Vec<Symbol>,
+}
+
+impl Annotations {
+    // This is limited to crate visibility to allow us to change `Vec<_>` to something
+    // else (`thinvec`?) in the future.
+    pub(crate) fn new(symbols: Vec<Symbol>) -> Self {
+        Annotations { symbols }
+    }
+
+    // Constructs an Annotations object representing an empty symbol sequence
+    pub(crate) fn empty() -> Self {
+        Annotations {
+            symbols: Vec::new(),
+        }
+    }
+
+    /// Returns a [SymbolsIterator] that yields each of the [Symbol]s in this annotations sequence
+    /// in order.
+    pub fn iter(&self) -> SymbolsIterator {
+        self.into_iter()
+    }
+
+    /// Returns the number of annotations in this sequence.
+    /// ```
+    /// use ion_rs::element::Annotations;
+    /// let annotations: Annotations = ["foo", "bar", "baz"].into();
+    /// assert_eq!(annotations.len(), 3);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.symbols.len()
+    }
+
+    /// Returns `true` if this sequence contains zero annotations. Otherwise, returns `false`.
+    /// ```
+    /// use ion_rs::element::Annotations;
+    /// let annotations: Annotations = ["foo", "bar", "baz"].into();
+    /// assert!(!annotations.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.symbols.len() == 0
+    }
+
+    /// Returns `true` if any symbol in this annotations sequence is equal to the provided text.
+    /// Otherwise, returns `false`.
+    /// ```
+    /// use ion_rs::element::Annotations;
+    /// let annotations: Annotations = ["foo", "bar", "baz"].into();
+    /// assert!(annotations.contains("foo"));
+    /// assert!(annotations.contains("bar"));
+    /// assert!(annotations.contains("baz"));
+    ///
+    /// assert!(!annotations.contains("quux"));
+    /// assert!(!annotations.contains("quuz"));
+    /// ```
+    pub fn contains<S: AsRef<str>>(&self, query: S) -> bool {
+        let query: &str = query.as_ref();
+        self.iter().any(|symbol| symbol.text() == Some(query))
+    }
+
+    /// Determines whether this annotations sequence begins with the specified sequence of text
+    /// values.
+    /// ```
+    /// use ion_rs::element::Annotations;
+    /// let annotations: Annotations = ["foo", "bar", "baz"].into();
+    /// assert!(annotations.starts_with("foo"));
+    /// assert!(!annotations.starts_with("quux"));
+    /// ```
+    pub fn starts_with(&self, text: impl AsRef<str>) -> bool {
+        self.iter().next().and_then(|a| a.text()) == Some(text.as_ref())
+    }
```

Comment:
> I agree that inspecting the first annotation is probably the most common thing to do with the annotations. However, I think people might want to be able to match on the first annotation. It is possible, but very unwieldy to match using a boolean function [....]

That's a good thought! I can get behind `first()`. 👍 

> By forcing the result to be a bool, I think that there is the slight possibility of users introducing subtle bugs when there is unknown symbol text. I say "slight" because I think unknown symbol text is uncommon in real world data, but if you do have unknown symbol text, it would be very easy to accidentally do the wrong thing with this function. Really, there are three possible outcomes—true, false, and indeterminate (in the case of unknown symbol text).

`ion-rust` makes a conscious decision to (at the user level) treat symbols whose text is unknown due to missing context as a hard error. This is different approach from `ion-java`, where it's possible to forge ahead in the absence of a shared symbol table if you do so with great caution.

`ion-rust` _cannot_, however, treat `$0` as an error as it is explicitly part of the domain of the `symbol` type. (i.e. Its text is _known to be unknown_, it's not unknown due to circumstance.)

By accepting an `AsRef<str>`, these convenience APIs are choosing not to support the `$0` case. (We could add support for it later with some trait work if we wanted to.) If the user asks "Is the first annotation `dog`?", the answer is `true` or `false`. If the first annotation is `$0`, it's not equivalent to the symbol 'dog'.

Reply:
> ion-rust cannot, however, treat $0 as an error as it is explicitly part of the domain of the symbol type. (i.e. Its text is known to be unknown, it's not unknown due to circumstance.)

This is one of the little corners of Ion that still makes my head spin sometimes, but you've helped me understand why `starts_with` and `contains` are not problematic.